hummingbot/client/command/gateway_command.py

```diff
@@ -102,6 +200,32 @@ async def _generate_certs(self,  # type: HummingbotApplication
         self.app.hide_input = False
         self.app.change_prompt(prompt=">>> ")
 
+    async def _generate_gateway_confs(self):
+
+        GENERATE_CONF_SCRIPT = path.join(root_path(), "gateway/setup/generate_conf.sh")
+
+        GATEWAY_CONF_PATH = self._get_gateway_config_path()
+        ETH_CONF_FILENAME = f"{GATEWAY_CONF_PATH}/ethereum.yml"
+
+        # Generates all the necessary gateway conf
+        subprocess.check_call([GENERATE_CONF_SCRIPT, self._get_gateway_config_path()],
+                              stdout=subprocess.DEVNULL)
+        self.notify(f"Gateway Configuration files are create in {GATEWAY_CONF_PATH}.")
+
+        self.app.clear_input()
+        self.placeholder_mode = True
+        self.app.hide_input = True
+
+        node_api_key: str = await self.app.prompt(prompt="Enter Infura API Key (required for Ethereum node, if you do not have one, make an account at infura.io):  >>> ")
+
+        ethereum_conf = await load_yml_into_dict(yml_path=ETH_CONF_FILENAME)
+        ethereum_conf["nodeAPIKey"] = node_api_key
+        await save_yml_from_dict(yml_path=ETH_CONF_FILENAME, conf_dict=ethereum_conf)
```

Comment:
This looks like a good idea, but it has edge cases that won't work.

If you're running Hummingbot from source in Linux, the UID and GID of files mounted to containers are the same as those on the host system.

Let's say Hummingbot's user is UID 1002, which is a common UID in Linux ... it'll first initialize gateway with config files of UID 1002, and then gateway will chown them to 8211.

Let's say Hummingbot wants to do a save_yml() call later. It'll fail, because it cannot write to a file owned by UID 8211.

The problem doesn't seem to happen on macOS due to how macOS docker does filesystem UID / GID mapping. But on Linux it'll break.

This is why I recommended you to use an initialization script on the gateway side instead - that method respects the permission boundaries between Hummingbot and gateway. Crossing the boundaries are always risky and prone to introduce maintenance issues later on.

Reply:
@martinkou I believe I understand your concern here. But if I have an initialization script that executes in the gateway container, the files created would still be owned by UID 8211 no? In which case the same " issue will occur when attempting to edit the configs from the host Linux.

Reply:
The gateway docker's user (8211) owns its config files - that's correct and is expected. That's in the same spirit as how you should not be able to edit your Linux system's NGINX or OpenSSH server configs without sudo.

Reply:
The important thing to note here is when you run anything with Docker, it's effectively got root permissions and its permissions boundaries would be different from running from source.

So, if you run gateway from source, it runs as your local user ID - then to you it looks like there's no permission boundary. But there is actually a permission boundary that other users should not be able to read or write to your home directory.

If you run gateway via Docker, then it should have its own user ID, just like you're running an HTTP server or email server in a Linux system. It'll have its own permission boundaries, and you should not be able to write to its configurations as your local user without sudo.

So you should not confuse the dev case (where Hummingbot and gateway are run as source and under the same permission boundary) and the deployment case (where Hummingbot and gateway are run as separate Dockers).